featuretools/entityset/entityset.py

```diff
@@ -410,52 +410,64 @@ def find_path(self, start_entity_id, goal_entity_id,
+        entity. Does not include paths which contain cycles.
 
         Args:
             start_entity_id (str) : id of entity to start the search from
             goal_entity_id  (str) : if of entity to find forward path to
 
-        Returns:
-            List of relationships that go from start entity to goal
-                entity. None is return if no path exists
-
         See Also:
-            :func:`BaseEntitySet.find_backward_path`
+            :func:`BaseEntitySet.find_backward_paths`
             :func:`BaseEntitySet.find_path`
         """
+        for sub_entity_id, path in self._forward_entity_paths(start_entity_id):
+            if sub_entity_id == goal_entity_id:
+                yield path
 
-        if start_entity_id == goal_entity_id:
-            return []
-
-        for r in self.get_forward_relationships(start_entity_id):
-            new_path = self.find_forward_path(
-                r.parent_entity.id, goal_entity_id)
-            if new_path is not None:
-                return [r] + new_path
-
-        return None
-
-    def find_backward_path(self, start_entity_id, goal_entity_id):
-        """Find a backward path between a start and goal entity
+    def find_backward_paths(self, start_entity_id, goal_entity_id):
+        """
+        Generator which yields all backward paths between a start and goal
+        entity. Does not include paths which contain cycles.
 
         Args:
             start_entity_id (str) : Id of entity to start the search from.
             goal_entity_id  (str) : Id of entity to find backward path to.
 
         See Also:
-            :func:`BaseEntitySet.find_forward_path`
+            :func:`BaseEntitySet.find_forward_paths`
             :func:`BaseEntitySet.find_path`
+        """
+        for path in self.find_forward_paths(goal_entity_id, start_entity_id):
+            # Reverse path
+            yield path[::-1]
 
-        Returns:
-            List of relationship that go from start entity to goal entity. None
-            is returned if no path exists.
+    def _forward_entity_paths(self, start_entity_id, seen_entities=None):
+        """
+        Generator which yields the ids of all entities connected through forward
+        relationships, and the path taken to each. An entity will be yielded
+        multiple times if there are multiple paths to it.
+
+        Implemented using depth first search.
         """
-        forward_path = self.find_forward_path(goal_entity_id, start_entity_id)
-        if forward_path is not None:
-            return forward_path[::-1]
-        return None
+        if seen_entities is None:
+            seen_entities = set()
+
+        if start_entity_id in seen_entities:
+            return
+
+        seen_entities.add(start_entity_id)
+
+        yield start_entity_id, []
+
+        for relationship in self.get_forward_relationships(start_entity_id):
+            next = relationship.parent_entity.id
```

Comment:
stylistic change, let's avoid overwriting the builtin `next` function. perhaps name it `next_relationship`?